### ISSUE TITLE
Auth Failures (echo chamber)

### ISSUE THREAD

**Issue by LP560** [DIGITALCRIMINAL/OnlyFans#341]
it seems OF has removed auth_hash when you inspect. is there any other way to find it? Or login

**Comment by WellSomething** [comment]
Yeah, scrapper has started to fail Auth Attempt, went to recheck the cookie information to see if everything was in order, and couldn't find auth_hash either. Not sure if anything else changed as well.

**Comment by artyom2035** [comment]
Getting failed auth attempts here too. Using chrome, I still see auth hash but mine has changed along with sess. Updated everything and still getting failed auths.

**Comment by jeremfg** [comment]
I still see the auth_hash cookie, but I noticed that the cookie "sess" is repeated in request header "access-token".  Perhaps that means something?

![image](https://user-images.githubusercontent.com/1882386/88482067-ea9eae80-cf2c-11ea-814f-30ca28ec3185.png)

In any case, I cannot authenticate anymore, even though all settings match between browser and config file. This is broken for me too.

**Comment by jeremfg** [comment]
I still see the auth_hash cookie, but I think it's because it's old. When I open an icognito window, I dont' see it anymore. However I see a cookie named "sc_is_visitor_unique" that I do not see in a normal window...

![image](https://user-images.githubusercontent.com/1882386/88482316-7c5aeb80-cf2e-11ea-8247-03e2241aae86.png)

**Comment by Hashirama** [comment]
I believe Onlyfans is adding two headers using XHR called "sign" and "time". If the request header doesn't contain these two you will get an error. It isn't a cookie error. They're both generated using javascript too.

**Comment by shinji257** [comment]
I've been seeing this issue as well.  As mentioned above one of the cookies is missing on new sessions but older ones still use it.

**Comment by BoxFit3** [comment]
> I believe Onlyfans is adding two headers using XHR called "sign" and "time". If the request header doesn't contain these two you will get an error. It isn't a cookie error. They're both generated using javascript too.

Is this issue easily fixable?

**Comment by DIGITALCRIMINAL** [comment]
They implemented this
https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html

**Comment by DIGITALCRIMINAL** [comment]
Depends, but we have to reverse engineer the encryption methods. I'm somewhat making progress, but I don't know if it will work.

**Comment by BoxFit3** [comment]
> Depends, but we have to reverse engineer the encryption methods. I'm somewhat making progress, but I don't know if it will work.

Damn, good luck bro, you the goat fr fr

**Comment by mwald84** [comment]
Not sure if this helps (or if you know about it already) but just in case...

https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html

**Comment by mwald84** [comment]
They have examples using Python there too (but you probably know about it already - I'm posting just in case):

https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html

Let us know if we can provide more data to help somehow.

**Comment by jumoog** [comment]
its just a SHA1 hash of 
accessToken from local storage + time + url + user agent + onlyfans
![grafik](https://user-images.githubusercontent.com/5268791/88492130-d7c9bf80-cfa8-11ea-9a77-7066be6c77d0.png)

**Comment by DIGITALCRIMINAL** [comment]
@jumoog, yee it is. I didn't want to say anything yet because I didn't want them changing anything lmao. They're probably watching this thread.

**Comment by jumoog** [comment]
good point i deleted it

**Comment by DIGITALCRIMINAL** [comment]
@jumoog have you tried sending a request with the signing and time?

**Comment by mileyvirus** [comment]
I tried sending a request with `sign` and `time` (and the usual stuff) and it seems to work for that specific endpoint, even multiple times, for quite a long time. I'm gonna do some more tests this weekend.

**Comment by mileyvirus** [comment]
Since it's a SPA and I see a CSRF in the cookies, the AWS keys to generate the signature are probably sloppily obfuscated in vendor.js.

**Comment by jumoog** [comment]
no its in the response for https://onlyfans.com/api2/v2/init?app-token=33d57ade8c02dbc5a333db99ff9ae26a

**Comment by jeremfg** [comment]
> @jumoog, yee it is. I didn't want to say anything yet because I didn't want them changing anything lmao. They're probably watching this thread.

lol... It's not like the code source is available anyway 😉. And they'll known it was "figured out" if the project is still active.

**Comment by DIGITALCRIMINAL** [comment]
> > @jumoog, yee it is. I didn't want to say anything yet because I didn't want them changing anything lmao. They're probably watching this thread.
> 
> lol... It's not like the code source isn't available anyway 😉. And they'll known it was "figured out" if the project is still active.

Yeah, I know, but I'd hate for them to change something whilst we're trying to figure it out.

**Comment by DIGITALCRIMINAL** [comment]
Yay
![image](https://user-images.githubusercontent.com/21285176/88500231-0010e880-cfc0-11ea-8254-14207e24a435.png)

**Comment by DIGITALCRIMINAL** [comment]
Alright so basically...

No 2FA = no auth_hash
2FA = auth_hash and auth_uniqu

You don't need the CSRF from what I can see.
Sess still needs to be passed. 

Regarding the new auth headers
You can pass any string for "sign"
You can pass any int for "time"

As @jumoog said (which was deleted) the signing is made up of the following things

sess_token
epoch timestamp
URL path (queries included)
user_agent
"onlyfans"

```
an69hkjp7kj169o691ugaiaqpj
3496797795845
/api2/v2/init?app-token=33d57ade8c02dbc5a333db99ff9ae26a
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36
onlyfans
```

You need to encode those values into sha1 encryption
https://emn178.github.io/online-tools/sha1.html

I'll make the script recreate the correct signing token since OF will know if you're using an invalid token.
They're probably allowing requests with the invalid sign and time values as a way to track users lmao, so I'll try my best to make it as anon as possible...

**Comment by DIGITALCRIMINAL** [comment]
```
def create_sign(session, link, sess, user_agent, text="onlyfans"):
    time = str(int(time2.time()))
    path = urlparse(link).path
    query = urlparse(link).query
    path = path+"?"+query
    a = [sess, time, path, user_agent, text]
    msg = "\n".join(a)
    message = msg.encode("utf-8")
    hash_object = hashlib.sha1(message)
    sha_1 = hash_object.hexdigest()
    session.headers["sign"] = sha_1
    session.headers["time"] = time
    return session
```

Created this function if anyone else wants to use it. It'll create the signed token and add it to the session's header.

**Comment by anonCoder123** [comment]
@DIGITALCRIMINAL i think there is a bug in the most recent push to master. I'm getting


```2020-07-27 02:37:26,573 ERROR errors invalid literal for int() with base 10: ''
Traceback (most recent call last):
  File "Z:\github-folders\OnlyFans\datascraper\main_datascraper.py", line 221, in start_datascraper
    result = x.start_datascraper(
  File "Z:\github-folders\OnlyFans\modules\onlyfans.py", line 92, in start_datascraper
    results = prepare_scraper(
  File "Z:\github-folders\OnlyFans\modules\onlyfans.py", line 519, in prepare_scraper
    export_archive(metadata_set, archive_directory, json_settings)
  File "Z:\github-folders\OnlyFans\helpers\main_helper.py", line 104, in export_archive
    datas2 = ofrenamer.start(archive_path, json_settings)
  File "Z:\github-folders\OnlyFans\extras\OFRenamer\start.py", line 90, in start
    metadata.valid = fix_metadata(
  File "Z:\github-folders\OnlyFans\extras\OFRenamer\start.py", line 50, in fix_metadata
    reformat = prepare_reformat(model2)
  File "Z:\github-folders\OnlyFans\extras\OFRenamer\start.py", line 47, in __init__
    self.maximum_length = int(text_length)
ValueError: invalid literal for int() with base 10: ''
```

**Comment by DIGITALCRIMINAL** [comment]
a new issue, please...

**Comment by anonCoder123** [comment]
@DIGITALCRIMINAL sure i can do that, i mentioned it here since it seems like it was a side effect of this fix.

**Comment by atliax** [comment]
Should I the auth_hash in the config be left blank with this "new" method? I don't know if I am doing something wrong, but I am still unable to log in with the script. I am doing everything else exactly the same way as before.

**Comment by NotMyMainOne** [comment]
Same issue as atliax. There's no auth hash in the cookies and I'm not sure what else needs to be filled in. Is 2FA mandatory now?

**Comment by DIGITALCRIMINAL** [comment]
If you see auth_hash in the cookies list, then yeah you gotta put it in the config

**Comment by NotMyMainOne** [comment]
I don't see auth_hash anymore. Now there's CSRF and sc_is_visitor_unique.

**Comment by DIGITALCRIMINAL** [comment]
Everything is the same as last time. Make sure you're using the correct user agent

**Comment by atliax** [comment]
My cookies are: sess, csrf, ref_src=; fp, auth_id, sc_is_visitor_unique
In my config I filled in: auth_id, sess, fp and user_agent

**Comment by BoxFit3** [comment]
Yeah I'm still having the same issue, filled it out correctly (auth_id, auth_hash_, sess, user_agent) and still having the Auth Attempt error

**Comment by shinji257** [comment]
I pulled down the new commit and it is working here.  Copied over my old config.json too.
Field populated are auth_id, auth_hash, sess, fp, and user_agent to match.  I also explicitly set support_2fa to false if it makes a difference at all since I don't have that enabled and therefore have no auth_uniq_ field to provide.

![image](https://user-images.githubusercontent.com/261005/88541219-627ae080-cfe2-11ea-8442-ffc48b9b63b4.png)

**Comment by Dalla-maker** [comment]
I see  auth_hash . And not  CSRF and sc_is_visitor_unique.

**Comment by AlexDucks** [comment]
I couldn't get it working at first when I tried to add updated values to my configuration file, even after pulling from master, but it did work after I cleared my cookies in my browser and logged back into OnlyFans. Here's what I did practically:

1. Cleared cookies for `onlyfans.com` (using developer tools).
	- In Brave (and Chrome probably):
		1. Open OnlyFans.com in my browser
		2. Hit F12 or CTRL+Shift+I for Developer Tools
		3. `Application` tab
		4. Cookies
		5. Right-click on `https://onlyfans.com/` and `Clear` ([Screenshot](https://i.ibb.co/HFQNt52/OFCookies-Brave.png))
	- Firefox Developer Tools are pretty similar, except:
		- Step 3: `Storage` tab, instead of application tab.
		- Step 5: Right-click on the _cookie values_ instead (not the URL) and click `Delete All from ".onlyfans.com"`. [Screenshot](https://i.ibb.co/bHLDkL0/OFCookies-Firefox.png)
	- Of course, an alternative is to just clear _all_ of your cookies, but it might be a bit overkill :)
2. Log back in as normal
3. Fill in new cookie values, where I had the following:
	- `auth_uniq_XXXXXXX` - Looks like it only shows up if you have 2FA enabled and you check the "Save this computer for 30 days" option.
	- `auth_id` - Unchanged
	- `sess`
	- `fp`

From what I can tell, `csrf` or any other value in cookies are not necessary. At least not for the time being.

I no longer had the `auth_hash` cookie value **at all**, so I just entered that as a blank string in the JSON config file - See short example:

```
...
"auth_id": "XXXXX",
"auth_hash": "",
"auth_uniq_": "YYYYYYYYYYYYYY",
...
```

**Comment by BatmanAndWilly** [comment]
> I no longer had the `auth_hash` cookie value **at all**

Same here, and it is working.

**Comment by atliax** [comment]
Interesting... I tried the latest commit again, doing everything exactly the same as yesterday and now it works perfectly.

The only difference I can think of is that yesterday the old version had been running recently, hammering the OF servers with "invalid sign" requests for a few hours. Maybe that's what caused my problems with the new version.

**Comment by DIGITALCRIMINAL** [comment]
OnlyFans stopped checking signed requests...
I'm still going to keep it there though.

**Comment by jumoog** [comment]
OnlyFans API is a fuckin joke

**Comment by BoxFit3** [comment]
Is it down again? Just tried it and getting the same error. Didn't change anything, tried restarting my machine, clearing cookies and still the same issue

**Comment by DIGITALCRIMINAL** [comment]
**It's been blocked again. Time for another sleepless night.
![image](https://user-images.githubusercontent.com/21285176/88686815-634d6880-d0ef-11ea-8f53-9bde3a38acc0.png)**

**Comment by BoxFit3** [comment]
Damn, take your time broski, inspect element ftw XD

**Comment by Hashirama** [comment]
It works fine for me? Or did they unblock it again?

**Comment by DIGITALCRIMINAL** [comment]
@Hashirama @Anonymeowse  fixed it in the latest commit

**Comment by mwald84** [comment]
To me the latest commit seems to authenticate (it shows my name) but then I get this error: 
2020-07-28 15:50:04,143 ERROR errors 'access-token'
Traceback (most recent call last):
  File "d:\OneDrive\cb\grab_of5\datascraper\main_datascraper.py", line 120, in start_datascraper
    session["session"], app_token, session["subscriber_count"], me_api, auth_count)
  File "d:\OneDrive\cb\grab_of5\modules\onlyfans.py", line 783, in get_subscriptions
    offset_array, [session]))
  File "D:\mrs\WPy64-3740\python-3.7.4.amd64\lib\multiprocessing\pool.py", line 276, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "D:\mrs\WPy64-3740\python-3.7.4.amd64\lib\multiprocessing\pool.py", line 657, in get
    raise self._value
  File "D:\mrs\WPy64-3740\python-3.7.4.amd64\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "D:\mrs\WPy64-3740\python-3.7.4.amd64\lib\multiprocessing\pool.py", line 47, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "d:\OneDrive\cb\grab_of5\modules\onlyfans.py", line 769, in multi
    r = json_request(session, link)
  File "d:\OneDrive\cb\grab_of5\helpers\main_helper.py", line 219, in json_request
    if session.headers["access-token"]:
  File "D:\mrs\WPy64-3740\python-3.7.4.amd64\lib\site-packages\requests\structures.py", line 52, in __getitem__
    return self._store[key.lower()][1]
KeyError: 'access-token'

I'a a creator so sometimes I have different errors. I'm also not using 2FA (don't know if useul info). I'm using same conf.json as before except I removed auth_hash

**Comment by DIGITALCRIMINAL** [comment]
Ye, it's a performer issue.
It's because the script creates an entirely new request.Session() when getting subscribers as a performer.
It'll fix it.

**Comment by DIGITALCRIMINAL** [comment]
@mwald84 I think the latest commit fixes it. It should include your username and others when getting subscribers.
If it's not fixed, I'll just login to creator account.

**Comment by mwald84** [comment]
Using the last commit:

2020-07-28 16:20:53,439 ERROR errors 'subscribedBy'
Traceback (most recent call last):
  File "d:\OneDrive\cb\grab_of5\datascraper\main_datascraper.py", line 120, in start_datascraper
    session["session"], app_token, session["subscriber_count"], me_api, auth_count)
  File "d:\OneDrive\cb\grab_of5\modules\onlyfans.py", line 811, in get_subscriptions
    subscribedBy = result["subscribedBy"]
KeyError: 'subscribedBy'

**Comment by DIGITALCRIMINAL** [comment]
Should work now with the latest commit

**Comment by mwald84** [comment]
It works! Thanks!

I know this is not the right place for this but do you have any diretions where I can understand better the new authentication system (funcions inside the onlyfans.py module maybe). I made a very simple script to check for new posts from a speciic model and send me a Telegram message. I only need to access the JSON file with the postsCount field. I'm checking your new code but I'm not sure I need to do all that only to have that files (I don't need to scrap anything else). Do you have a forum where I can post this kind of questions?

**Comment by mwald84** [comment]
I know this is not the proper place or this (please let me know if there's a forum for questions) but
do you think they check IP during authentication? I made a simple script to only authenticate, check the number of posts from a speciic model and print on screen. I used to run this on a remote server. Now, with the new authentication, I can run the script on my computer (same I used the browser) but it can't autenticate on the remote server.

**Comment by avekifes** [comment]
2020-07-28 18:57:27,214 ERROR errors string indices must be integers
Traceback (most recent call last):
File "C:\Users\username\Desktop\OnlyFans-master\datascraper\main_datascraper.py", line 119, in start_datascraper
array = x.get_subscriptions(
File "C:\Users\username\Desktop\OnlyFans-master\modules\onlyfans.py", line 796, in get_subscriptions
identifier = result["username"]
TypeError: string indices must be integers

Is anyone else getting this, or did I screw something up in .settings?

**Comment by dannylaroos** [comment]
i've been trying to get it to work for ages now and despite filling in all the required information, i either get:

Traceback (most recent call last):
  File "J:\Python\OnlyFans-5.1\datascraper\main_datascraper.py", line 92, in sta
rt_datascraper
    app_token = json_auth['app_token']
KeyError: 'app_token'


or i get:

Auth (V1)
Auth Attempt 1/10
Auth Attempt 2/10
Auth Attempt 3/10
Auth Attempt 4/10
Auth Attempt 5/10
Auth Attempt 6/10
Auth Attempt 7/10
Auth Attempt 8/10
Auth Attempt 9/10
Auth Attempt 10/10
Auth (V2)
Auth Attempt 1/10
Auth Attempt 2/10
Auth Attempt 3/10
Auth Attempt 4/10
Auth Attempt 5/10
Auth Attempt 6/10
Auth Attempt 7/10
Auth Attempt 8/10
Auth Attempt 9/10
Auth Attempt 10/10
There's nothing to scrape.


have i missed anything?

**Comment by Anonymeowse** [comment]
@dannylaroos Is `app_token` set in your config.json alongside the fields you listed in your comment?

**Comment by avekifes** [comment]
> 
> 
> 2020-07-28 18:57:27,214 ERROR errors string indices must be integers
> Traceback (most recent call last):
> File "C:\Users\username\Desktop\OnlyFans-master\datascraper\main_datascraper.py", line 119, in start_datascraper
> array = x.get_subscriptions(
> File "C:\Users\username\Desktop\OnlyFans-master\modules\onlyfans.py", line 796, in get_subscriptions
> identifier = result["username"]
> TypeError: string indices must be integers
> 
> Is anyone else getting this, or did I screw something up in .settings?

Never mind; it's magically working again.

**Comment by zedawing** [comment]
> I couldn't get it working at first when I tried to add updated values to my configuration file, even after pulling from master, but it did work after I cleared my cookies in my browser and logged back into OnlyFans. Here's what I did practically:
> 
> 1. Cleared cookies for `onlyfans.com` (using developer tools).
>    
>    * In Brave (and Chrome probably):
>      
>      1. Open OnlyFans.com in my browser
>      2. Hit F12 or CTRL+Shift+I for Developer Tools
>      3. `Application` tab
>      4. Cookies
>      5. Right-click on `https://onlyfans.com/` and `Clear` ([Screenshot](https://i.ibb.co/HFQNt52/OFCookies-Brave.png))
>    * Firefox Developer Tools are pretty similar, except:
>      
>      * Step 3: `Storage` tab, instead of application tab.
>      * Step 5: Right-click on the _cookie values_ instead (not the URL) and click `Delete All from ".onlyfans.com"`. [Screenshot](https://i.ibb.co/bHLDkL0/OFCookies-Firefox.png)
>    * Of course, an alternative is to just clear _all_ of your cookies, but it might be a bit overkill :)
> 2. Log back in as normal
> 3. Fill in new cookie values, where I had the following:
>    
>    * `auth_uniq_XXXXXXX` - Looks like it only shows up if you have 2FA enabled and you check the "Save this computer for 30 days" option.
>    * `auth_id` - Unchanged
>    * `sess`
>    * `fp`
> 
> From what I can tell, `csrf` or any other value in cookies are not necessary. At least not for the time being.
> 
> I no longer had the `auth_hash` cookie value **at all**, so I just entered that as a blank string in the JSON config file - See short example:
> 
> ```
> ...
> "auth_id": "XXXXX",
> "auth_hash": "",
> "auth_uniq_": "YYYYYYYYYYYYYY",
> ...
> ```

This worked for me

**Comment by DIGITALCRIMINAL** [comment]
> @dannylaroos Is `app_token` set in your config.json alongside the fields you listed in your comment?

Download the latest commit, not 5.1

**Comment by avekifes** [comment]
> 
> 
> 2020-07-28 18:57:27,214 ERROR errors string indices must be integers
> Traceback (most recent call last):
> File "C:\Users\username\Desktop\OnlyFans-master\datascraper\main_datascraper.py", line 119, in start_datascraper
> array = x.get_subscriptions(
> File "C:\Users\username\Desktop\OnlyFans-master\modules\onlyfans.py", line 796, in get_subscriptions
> identifier = result["username"]
> TypeError: string indices must be integers
> 
> Is anyone else getting this, or did I screw something up in .settings?

This error seems to be completely random. Sometimes it even shows up in the middle of a scrape.

**Comment by dannylaroos** [comment]
> @dannylaroos Is `app_token` set in your config.json alongside the fields you listed in your comment?

yes it is.  forgot to mention it first time round.

**Comment by anonCoder123** [comment]
hmm my stuff just randomly stopped working again.

**Comment by thepetcow** [comment]
I am having the same issue (Access Denied).

I've logged out and logged back in, cleared Cookies, I am only getting the "auth_id", "sess" "app-token" and "user_agent" from the cookie and I am copying and pasting these right from the DevTools.

I'm using the new commit from a few hours ago.

**Comment by lordyspam** [comment]
Hi.
I tried some days with de lastest version without results, always "Access Denied", i disable now 2FA and my config file is this:

`"supported": {
	"onlyfans": {
			"auth": {
				"username": "uXXXXX",
				"auth_id": "XXXXXX", 
				"sess": "XXXXXX",
				"app-token": "33d57ade8c02dbc5a333db99ff9ae26a",
				"auth_hash": "",
				"auth_uniq_": "",
				"fp": "XXXXXXX",
				"user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36"
		}
	}
}
`

is correct?
Thanks u!

**Comment by kervala** [comment]
That's strange because it worked fine yesterday and today it detects it's "OnlyFans (V1)" and auth fails with "Access Denied" :(

I checked several times all variables were correct in config file.

**Comment by kervala** [comment]
Nevermind, it still works, I'm sorry :) That was because Vivaldi is using a mobile user-agent while developer console is shown, so the user-agent was different :(

**Comment by minayondaim3** [comment]
How exactly are you supposed to get this working again? Yall kinda all over the place

**Comment by DIGITALCRIMINAL** [comment]
> How exactly are you supposed to get this working again? Yall kinda all over the place

It's already been fixed.

**Comment by atliax** [comment]
> How exactly are you supposed to get this working again? Yall kinda all over the place

@minayondaim3 At the moment, you need to download the latest commit (not the release from June 20th). That can be accessed here: https://github.com/DIGITALCRIMINAL/OnlyFans/archive/master.zip

Then just follow the "mandatory tutorial" part of the readme and it should work.

**Comment by Psiplexser** [comment]
I keep getting this error

C:\Users\\Desktop\Onlyfans\OnlyFans-master>start_ofd.py
Site: 0 = onlyfans | 1 = starsavn | 2 = fourchan | 3 = bbwchan | x = Exit

2020-08-05 20:46:06,958 ERROR errors invalid literal for int() with base 10: ''
Traceback (most recent call last):
  File "C:\Users\\Desktop\Onlyfans\OnlyFans-master\datascraper\main_datascraper.py", line 71, in start_datascraper
    x = int(x)
ValueError: invalid literal for int() with base 10: ''

**Comment by Psiplexser** [comment]
Keep getting access denied. I can find auth_uniq but not auth hash is that why?

cookies i can see are Auth_id, auth_uniq, sess, csrf, fp, streams, st

**Comment by secretiveolwagner** [comment]
Please see @zedawing post ofrom [above](https://github.com/DIGITALCRIMINAL/OnlyFans/issues/341#issuecomment-665415780), the method is working me

**Comment by zorzon** [comment]
hey folks just downloaded the master this evening and spent like 5 hours trying to get this to work.

i've tried with 2FA and without, ( the auth_hash hasn't appeared any more since the first time i tried it, without 2FA)
and resetting cookies between attempts as demonstrated above.

i input auth_id, sess, user_agent, and app token (already there), and fp:
some combos produce a stream of auth attempt access denied, but when i've reset everything and started over following the README once again it produces a 
**"Welcome _username_ (hurray)
"There's nothing to scrape." (f@%k)"**

my question is as follows, as i'm pretty new to python and scraping etc.:
am i doing something wrong?! if there's nothing to scrape am i on the wrong page maybe?
i hit *0* and then *Enter* while i have the session open in the Chrome window, with the page open at onlyfans[.]com/amodelsusername lol. what i see in my window is the account header and bio, profile pic etc, and posts with a button saying subscribe for $420.69 or whatever. that's when i hit *0* and then *Enter* and produce the "nothing to scrape" dialog.

am i a moron? is this a program for people who have already subbed? just to download available content?
i thought this would be able to download content without having to pay, so if i'm mistaken, or if this no longer works, or if anyone has any ideas or help essentially for a n00b i'd hugely appreciate. i really feel like i'm messing something up

thank you so much in advance
z
![computer-trouble-3579189](https://user-images.githubusercontent.com/69563125/89984263-c60f3a00-dc81-11ea-8a82-6379b87e992b.jpg)

**Comment by DIGITALCRIMINAL** [comment]
> am i a moron? is this a program for people who have already subbed? just to download available content?
> i thought this would be able to download content without having to pay, so if i'm mistaken, or if this no longer works, or if anyone has any ideas or help essentially for a n00b i'd hugely appreciate. i really feel like i'm messing something up

I'm sorry friend, but yes you are a moron.
Well tbh, you came late. There was an unreproducible bug in my script which allowed people to auth into other user's account earlier this year. We scraped all the content and dipped.

**Comment by zorzon** [comment]
okey great! lol thanks for filling me in. i thought this was recent and working because the last comment was from 5d ago. anyway i learnt a bit more about python in the process so i'll take that as a small win

any advice where to keep an eye out if anything else like this ever crops up?

if not no worries thanks again 🙌

**Comment by jumbled37** [comment]
> okey great! lol thanks for filling me in. i thought this was recent and working because the last comment was from 5d ago. anyway i learnt a bit more about python in the process so i'll take that as a small win
> 
> any advice where to keep an eye out if anything else like this ever crops up in the future?
> 
> if not no worries thanks again 🙌

You are correct It’s recent and working when you set it up correctly, which in fairness it does seem you have done, you just need to do the final and most important part of getting this to work and that is

**Comment by simonk718** [comment]
it's currently not working for me, I followed all advice from this thread. I had it working before.

**Comment by SteveJSzabo** [comment]
This was working for me until two days ago. Now I'm not able to authenticate.

**Comment by Psiplexser** [comment]
Does this work with Manyvids?

**Comment by iDerp** [comment]
I could be wrong, but it looks like there is a whole new issue. Auth doesn't work at all now, and from some tests it may be due to a missing header. Ive also noticed the cookie it uses doesn't seem to ever need to be updated, so why not always use the same cookie that could be provided in config? I just don't see a reason to reauth every time you use the program, because the old auth still works. I tried to fix this myself but I'm not a python dev so I didn't get very far.

**Comment by DIGITALCRIMINAL** [comment]
@iDerp ? You don't need to reauth every time? I'm still using my month old cookie settings.

**Comment by iDerp** [comment]
Well I could be wrong again, since im no python dev, but in the code it looks like now when calling the customer endpoint, or any really, it makes its own cookie. What I mean is that it doesn't seem to need to do that. I'm also getting `Access Denied.` when it tries to login all of a sudden, even after getting a new auth for it. But the main issue here probably isn't the cookie, but instead missing headers, as when I did my own test I got the same error, until I added all the headers back in (and it didnt even seem to need the sign or time header either)

**Comment by DIGITALCRIMINAL** [comment]
We need to send a sign and time to send a valid request to the endpoint.

If you're missing a header, it's best you download Postman and figure which header you're missing because I wouldn't know.

**Comment by iDerp** [comment]
Good thing I already use Postman lol. Alright yeah after some tests it seems if you dont update time and sign header, they expire. (Didn't know that lol) Its weird, it works sometimes, and then it just stops. I'll have to do more testing it seems to see if its an issue on my end.

**Comment by DIGITALCRIMINAL** [comment]
I haven't been getting auth failures that I've had to debug. There are certain timings the script uses. Are you using a model account?

**Comment by Psiplexser** [comment]
I just tried today and didn't run into any issues.

**Comment by iDerp** [comment]
Interesting, and it's just a regular account. Knowing me I might have screwed something up, but I don't recall doing anything wrong, who knows. I'll try it again when I can and see what happens.

**Comment by iDerp** [comment]
Still having the same issue. Everytime I try and start the script it tells me the same thing and I get logged out on the browser.

**Comment by Hashirama** [comment]
> Still having the same issue. Everytime I try and start the script it tells me the same thing and I get logged out on the browser.
> 
> Update: Looks like their either dont like my IP or browser anymore :/ sorry for the troubles

Are you using the correct user agent?

**Comment by iDerp** [comment]
I was, but then I tried a different browser, used the new user-agent, new sess and everything and it started working. So im not 100% what fixed it but something did lol

**Comment by SteveJSzabo** [comment]
I tried using a different browser (and agent, sess, etc) and still am unable to authenticate. Should I try using the authentication v2 only instead of starting with v1?

**Comment by danssdan** [comment]
It does not download any of the paid content even though I am a subscriber. Help?

**Comment by Zecfigh404** [comment]
i had the same problem, how to fix it?

**Comment by secretiveolwagner** [comment]
@Zecfigh404 I had the same problem when upgrading from 5.0 to 5.2. Basically copypasta all the credentials, I also made sure the my Onlyfans session is still logged in but it kept saying Access denied. I then cleared the cookies, re-logged in to Onlyfans and eventually made it to work. I 'll admit, the latest version seems to be buggier than previous versions, I hope @DIGITALCRIMINAL can take a look at this.

**Comment by Zecfigh404** [comment]
> @Zecfigh404 I had the same problem when upgrading from 5.0 to 5.2. Basically copypasta all the credentials, I also made sure the my Onlyfans session is still logged in but it kept saying Access denied. I then cleared the cookies, re-logged in to Onlyfans and eventually made it to work. I 'll admit, the latest version seems to be buggier than previous versions, I hope @DIGITALCRIMINAL can take a look at this.

so i have to downgrade? can u specific tell me what is the credentials? what should i write in auth_hash column?

**Comment by secretiveolwagner** [comment]
> 
> 
> > @Zecfigh404 I had the same problem when upgrading from 5.0 to 5.2. Basically copypasta all the credentials, I also made sure the my Onlyfans session is still logged in but it kept saying Access denied. I then cleared the cookies, re-logged in to Onlyfans and eventually made it to work. I 'll admit, the latest version seems to be buggier than previous versions, I hope @DIGITALCRIMINAL can take a look at this.
> 
> so i have to downgrade? can u specific tell me what is the credentials? what should i write in auth_hash column?

@Zecfigh404 you'll find your answer from above.

**Comment by bootyislife** [comment]
Yeah it stopped authing. Tried the ways mentioned above but nothing. Even with the newest release.

**Comment by kennhardy** [comment]
Just cloned the latest version. Does not seem to work. Even with above method as well.

**Comment by secretiveolwagner** [comment]
@bootyislife @kennhardy If you guys want some help, instead of just straight up saying the program is not working, at the very least tell us what your attempted solution(s). No one's gonna help you if you don't help yourself first.

**Comment by UncleVladd** [comment]
I'm confused wether I should be enabling 2fa or not to get the auth hash.
@secretiveolwagner, I know you said the answer is above ,but it gets really confusing for newcomers. 

So my question is:

do I downgrade the version?
do I use 2fa to get the hash ?

Thanks and sorry to newbie bomb the support channel.

**Comment by AlexDucks** [comment]
I am using the latest version from master and it works fine for me right now (even with 'fresh' credentials).

`auth_hash` is not required. Regardless of 2FA status. If you don't see `auth_hash` in your browser, just leave it as blank `"auth_hash": "",`.

If you have 2FA enabled, you need to include the `auth_uniq_` value from your browser. Without it, the application will ask for a 2FA code and entering a 2FA code should work fine too (I haven't tried that method in a few days though).

For those of you that says it doesn't work, make sure that **ALL** the values are correct and UP-TO-DATE. Your web browser might have auto-updated and in that case the `user_agent` field in the config.json file might be wrong.

I am currently using the latest version of Brave on Windows (at least I don't have an update right now) and my `user_agent` value reflects that. Obviously you will have to replace it with the one you have and make sure it **does not have any "spaces" before or after the text**.  
You can use a website like https://www.whatsmyua.info/ to make sure you're not copying any invisible extra characters you're not supposed to have in the `user_agent` value.

Here's what my `auth` section of the config file looks like, obviously fill in with values from your own browser cookies and such:

```json
"auth": {
    "username": "YOUR_USERNAME_HERE",
    "auth_id": "THIS_IS_ONLY_NUMBERS",
    "auth_hash": "",
    "auth_uniq_": "ONLY_IF_2FA_ENABLED",
    "sess": "REQUIRED_ALWAYS",
    "fp": "REQUIRED_ALWAYS",
    "app_token": "33d57ade8c02dbc5a333db99ff9ae26a",
    "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36",
    "support_2fa": true
},
```

**Comment by Hashirama** [comment]
Yeah he's right ^

You don't need the auth_hash. It works fine without, for now.

**Comment by UncleVladd** [comment]
@AlexDucks Thank you very much for clearing that up. Not only makes sense. It works perfectly!

**Comment by SteveJSzabo** [comment]
> I am using the latest version from [`master`](https://github.com/DIGITALCRIMINAL/OnlyFans/archive/master.zip) ('Download ZIP' option) and it works fine for me right now (even with 'fresh' credentials).
> 
> `auth_hash` is not required (as of right now at least). Regardless of 2FA status. If you don't see `auth_hash` in your browser, just leave it as blank `"auth_hash": "",`.
> 
> If you have 2FA enabled, you need to include the `auth_uniq_` value from your browser. Without it, the application will ask for a 2FA code and entering a 2FA code should work fine too (I haven't tried that method in a few days though).
> 
> For those of you that says it doesn't work, make sure that **ALL** the values are correct and UP-TO-DATE. Your web browser might have auto-updated and in that case the `user_agent` field in the config.json file might be wrong.
> 
> I am currently using the latest version of Brave on Windows (at least I don't have an update right now) and my `user_agent` value reflects that. Obviously you will have to replace it with the one you have and make sure it **does not have any "spaces" before or after the text**.
> You can use a website like https://www.whatsmyua.info/ to make sure you're not copying any invisible extra characters you're not supposed to have in the `user_agent` value.
> 
> Here's what my `auth` section of the config file looks like, obviously fill in with values from your own browser cookies and such:
> 
> ```json
> "auth": {
>     "username": "YOUR_USERNAME_HERE",
>     "auth_id": "THIS_IS_ONLY_NUMBERS",
>     "auth_hash": "",
>     "auth_uniq_": "ONLY_IF_2FA_ENABLED",
>     "sess": "REQUIRED_ALWAYS",
>     "fp": "REQUIRED_ALWAYS",
>     "app_token": "33d57ade8c02dbc5a333db99ff9ae26a",
>     "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36",
>     "support_2fa": true
> },
> ```

I think perhaps the user agent extra characters was doing it for me. I tried copying my user agent from the site you provided and voila! I'm up and running again. Thanks!

**Comment by secretiveolwagner** [comment]
@SteveJSzabo Speaking of user agent, the way I do it is to use a portable version of a browser, that way you'll get a consistent user agent because it never updates automatically.

**Comment by M-rcus** [comment]
I made a browser extension that makes it easier to copy the required values for `config.json` / `extra_auth.json`: https://github.com/M-rcus/OnlyFans-Cookie-Helper

As of right now it requires manual installation in your browser, but hopefully I'll be able to make it easier in the future.

The repository page for the extension should explain how it works and such, but feel free to contact me by mentioning me here, [emailing me: m@rcus.dev](mailto:m@rcus.dev).  
Creating an issue in my repository might be fine too, but I'd prefer if that's only for any bugs or issues unrelated to _installing_ the extension that you might stumble upon.

Shoutouts to @AlexDucks for giving me the idea for it and helping with testing :P

**Comment by secretiveolwagner** [comment]
@M-rcus Works perfectly, thanks!

**Comment by gerardvp** [comment]
Running PopOS and I could not get auth to work at all. Kept having to relogin and change my fp and sess. I used the site @AlexDucks mentioned for the User Agent and it was different than what my browser and cookies had. Different than what the Cookie Helper had as well.

**Browser and Cookies:** Mozilla/5.0 (X11; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0
**User Agent Website:** Mozilla/5.0 (Windows NT 10.0; rv:78.0) Gecko/20100101 Firefox/78.0

Using the User Agent Website one ended up getting my auth to work and finally login.
Sidenote: "Username" is not the username you signed up with.

**Comment by DIGITALCRIMINAL** [comment]
Username in the config is optional and doesn't do anything. It's just there to track which account you're authing into

**Comment by M-rcus** [comment]
> Running PopOS and I could not get auth to work at all. Kept having to relogin and change my fp and sess. I used the site @AlexDucks mentioned for the User Agent and it was different than what my browser and cookies had. Different than what the Cookie Helper had as well.
> 
> **Browser and Cookies:** Mozilla/5.0 (X11; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0
> **User Agent Website:** Mozilla/5.0 (Windows NT 10.0; rv:78.0) Gecko/20100101 Firefox/78.0
> 
> Using the User Agent Website one ended up getting my auth to work and finally login.

The cookie helper extension is pretty "lazy" and uses [`navigator.userAgent`](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorID/userAgent) in JavaScript to access the user agent, which is what the browser has internally.

I suppose you might be either using Firefox's `resistFingerprinting` option or an extension that changes your user agent for websites - in which case `navigator.userAgent` does report the wrong user agent.
I'll update the cookie helper extension to use an external API instead of relying on `navigator.userAgent`

> Sidenote: "Username" is not the username you signed up with.

As @DIGITALCRIMINAL said - the `username` field in the config does not matter. It's purely meant for easy tracking of the user.
The cookie helper extension is lazy here too, since all it does is combine the letter "u" with the `auth_id` field.

**Comment by omgcyrilic** [comment]
not working for me with "fresh" values ... "fp" wasn't required for me before but I filled it anyway. using the latest code from the repo.

is it working for everyone else right now?

**Comment by uslacker99** [comment]
Has OnlyFans changed something again? I seem to get the 
Auth Attempt 10/10
Access denied.

**Comment by DIGITALCRIMINAL** [comment]
Lul just got an email from someone saying that various scripts and extensions are getting copyright claimed by OnlyFans because we used the name "OnlyFans"

I'm probably next, so I've added some disclaimers to the readme.

**Comment by secretiveolwagner** [comment]
> Lul just got an email from someone saying that various scripts and extensions are getting copyright claimed by OnlyFans because we used the name "OnlyFans"
> 
> 
> 
> I'm probably next, so I've added some disclaimers to the readme.

Thanks for the heads up. Time to backup

**Comment by SteveJSzabo** [comment]
I'm now getting a 2FA prompt that when I enter the 2FA code, it still fails. I've been using the Cookie Helper extension here too. Is there a specific format the 2FA code needs?

**Comment by timothychare** [comment]
Firstly, I want to thank you guys for all of this :) you guys are the real MVPs

I keep getting access denied. I downloaded a fresh zip of the script, and pulled my cookie details from 3 different browsers and still no luck. Any logs I can check to see what might be failing exactly?

**Comment by DIGITALCRIMINAL** [comment]
> Firstly, I want to thank you guys for all of this :) you guys are the real MVPs
> 
> I keep getting access denied. I downloaded a fresh zip of the script, and pulled my cookie details from 3 different browsers and still no luck. Any logs I can check to see what might be failing exactly?

Try using the cookie helper

**Comment by leeharveyosmond** [comment]
> Firstly, I want to thank you guys for all of this :) you guys are the real MVPs
> 
> I keep getting access denied. I downloaded a fresh zip of the script, and pulled my cookie details from 3 different browsers and still no luck. Any logs I can check to see what might be failing exactly?

Same here.  Using Chrome on OSX 10.14.6 with Python 3.8 installed.
Tried to scrape, authentication failed, so I updated config.json with fresh cookies, and ... failed again.
I did 'git pull', 'git rev-parse HEAD' says fac67e247c76c7d244e030bb53ae4edaa67bf6f8.
That gave me a problem:
  File ".... /OnlyFans-master/helpers/main_helper.py", line 16, in <module>
    import psutil
ModuleNotFoundError: No module named 'psutil'
Fixed that; still can't get in.
In particular I notice that if I log out of OnlyFans in Chrome, then log back in, it appears to be still using the same cookie values successfully; I'm seeing lots of "Status Code: 200" in the headers/general bit without the  annotations that appear for cache hits.

**Comment by tehgarra** [comment]
> 
> 
> > Firstly, I want to thank you guys for all of this :) you guys are the real MVPs
> > I keep getting access denied. I downloaded a fresh zip of the script, and pulled my cookie details from 3 different browsers and still no luck. Any logs I can check to see what might be failing exactly?
> 
> Same here. Using Chrome on OSX 10.14.6 with Python 3.8 installed.
> Tried to scrape, authentication failed, so I updated config.json with fresh cookies, and ... failed again.
> I did 'git pull', 'git rev-parse HEAD' says [fac67e2](https://github.com/DIGITALCRIMINAL/OnlyFans/commit/fac67e247c76c7d244e030bb53ae4edaa67bf6f8).
> That gave me a problem:
> File ".... /OnlyFans-master/helpers/main_helper.py", line 16, in
> import psutil
> ModuleNotFoundError: No module named 'psutil'
> Fixed that; still can't get in.
> In particular I notice that if I log out of OnlyFans in Chrome, then log back in, it appears to be still using the same cookie values successfully; I'm seeing lots of "Status Code: 200" in the headers/general bit without the annotations that appear for cache hits.
> 
> EDIT well huh. Came back two hours later, tried again with Safari, got invited to do a recaptcha before proceeding with logging in, got in, [much harder to get at the cookies with Safari], now scraping.

how did you figure out the psutil issue? i'm having the same problem currently

**Comment by leeharveyosmond** [comment]
> how did you figure out the psutil issue? i'm having the same problem currently

on this page https://github.com/giampaolo/psutil/blob/master/INSTALL.rst
I found these magic spells:
wget https://bootstrap.pypa.io/get-pip.py -O - | python
python < <(curl -s https://bootstrap.pypa.io/get-pip.py)
and they did not work for me.
So I picked one of them apart (I chose the curl one) and examined what it downloaded.  Turns out, it was a python script starting with a big chunk of data.
So I ran than script with python3 [any Mac user that's got hal this far will know the install python installation won't cut it, need to download and install python3, need to be careful about $PATH and if you're a nutjob $DYLD_FRAMEWORK_PATH and $DYLD_LIBRARY_PATH too, but I digress] ... and that didn't cut it either.

I then found https://pypi.org/project/psutil/#files  , downloaded the appropriate package (for me this was psutil-5.7.2.tar.gz).  I then built that (unpack the tarball ; cd to the location ; then gnumake install), after that start_ofd.py ran cleanly and I was back facing the other issues I reported and eventually solved.

Do we need only the second item, or the first too once the second is done? Your guess is as good as mine.

**Comment by williamck123** [comment]
Any reason why it would stop working? Even when it has worked before? I checked the cookies and everything seems to match. Using version 5.6.1

**Comment by M-rcus** [comment]
> Any reason why it would stop working? Even when it has worked before? I checked the cookies and everything seems to match. Using version 5.6.1

Without knowing 100%, 5.6.1 is _relatively "old"_ and there might've been API changes on OnlyFans that caused that version to stop working.

It's probably best if you update to the latest 6.0 version, but keep in mind there are config changes that might need manual intervention*

\* I say that because I manually updated directly from the "development version" (master branch) and had some issues where my directory format wasn't 'converted' correctly.

**Comment by Hashirama** [comment]
> Any reason why it would stop working? Even when it has worked before? I checked the cookies and everything seems to match. Using version 5.6.1

You check your user agent?

**Comment by likefunbutnot** [comment]
It doesn't seem to be working for me, either. I get 

Auth (V1) Attempt 1/10
Auth (V1) Attempt 2/10
Auth (V1) Attempt 3/10
Auth (V1) Attempt 4/10
Auth (V1) Attempt 5/10
Auth (V1) Attempt 6/10
Auth (V1) Attempt 7/10
Auth (V1) Attempt 8/10
Auth (V1) Attempt 9/10
Auth (V1) Attempt 10/10
Scraping Paid Content
Scraping Subscriptions
There's nothing to scrape.
Archive Completed in 0.01 Minutes
Pausing scraper for 0 seconds.
Site: 0 = onlyfans | x = Exit

On a loop after providing all current cookie info from either Firefox or Chrome. I'm basing this off the info that SteveJSzabo  suggests is need and I am using a the script available on December 14, 2020 (the current date).

**Comment by williamck123** [comment]
Reinstalled to update to 6.0, put in new cookies, etc. Python now crashes pretty instantly when I launch start_ofd.py. Solutions?

**Comment by paisimisgontai** [comment]
+1 to https://github.com/DIGITALCRIMINAL/OnlyFans/issues/341#issuecomment-745039863

I'm getting the same results -- unable to auth with Chrome or Safari cookie, using the latest `.profiles/default/auth.json` file, etc.

My Network tab in the inspector is showing that I'm fetching from their v2 API -- and the scraper seems to only be hitting v1. Could this be part of the issue?

**Comment by rando-username** [comment]
I'm using the cookie copier extension and pulled latest master, getting the same as likefunbutnot.

**Comment by solipcism** [comment]
@williamck123 If you're getting the "nothing to scrape" errors, make sure you've added an `"active": true` property to your auth object.
The cookie copier doesn't add one, so the downloader will ignore the profile. It is noted in the readme.

Great utility Mx D. Crim.

**Comment by M-rcus** [comment]
> The cookie copier doesn't add one, so the downloader will ignore the profile. It is noted in the readme.

Older versions of the cookie helper extension didn't (because the `active` parameter wasn't a thing), but v1.0.2 - released December 26th, 2020 (so its been out for almost two months) - does: https://github.com/M-rcus/OnlyFans-Cookie-Helper/releases/tag/1.0.2

**Comment by j-ktz** [comment]
Getting Auth attempts repeating all of the sudden and then "nothing to scrape" with no changes made to my auth.json. It's worked up until now. Is this an OF issue?

Auth (V1) Attempt 1/10
Auth (V1) Attempt 2/10
Auth (V1) Attempt 3/10
Auth (V1) Attempt 4/10
Auth (V1) Attempt 5/10
Auth (V1) Attempt 6/10
Auth (V1) Attempt 7/10
Auth (V1) Attempt 8/10
Auth (V1) Attempt 9/10
Auth (V1) Attempt 10/10
Scraping Paid Content
Scraping Subscriptions
There's nothing to scrape.
